cudapoa/src/main.cpp

```diff
@@ -0,0 +1,277 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <claraparabricks/genomeworks/cudapoa/utils.hpp> // for get_multi_batch_sizes()
+#include "application_parameters.hpp"
+
+namespace claraparabricks
+{
+
+namespace genomeworks
+{
+
+namespace cudapoa
+{
+
+std::unique_ptr<Batch> initialize_batch(bool msa, bool banded_alignment, const double gpu_mem_allocation, const BatchSize& batch_size)
+{
+    // Get device information.
+    int32_t device_count = 0;
+    CGA_CU_CHECK_ERR(cudaGetDeviceCount(&device_count));
+    assert(device_count > 0);
+
+    size_t total = 0, free = 0;
+    cudaSetDevice(0); // Using first GPU for sample.
+    cudaMemGetInfo(&free, &total);
+
+    // Initialize internal logging framework.
+    Init();
+
+    // Initialize CUDAPOA batch object for batched processing of POAs on the GPU.
+    const int32_t device_id      = 0;
+    cudaStream_t stream          = 0;
+    size_t mem_per_batch         = gpu_mem_allocation * free; // Using 90% of GPU available memory for CUDAPOA batch.
+    const int32_t mismatch_score = -6, gap_score = -8, match_score = 8;
```

Comment:
I forgot about the scores, they may also need to be defined in the input, with the current default values.

Reply:
with checks about the sign. _gap_ and _mismatch_ should be negative, _match_ should be positive.

Reply:
Are 0s okay?

Reply:
as long as mismatch and gap scores are less than match score, it should be fine. Additionally, in some places in the code, such as `use32bitScore()`, there is an underlying assumption that match score is non-negative and gap/mismatch score are non-positive. So 0 should be OK.

Reply:
Done.